lib/recursive-methods.rb

```diff
@@ -1,49 +1,98 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    #base
+    return 1 if n == 0
+    raise ArgumentError if n < 0
+    #recursion
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n)
 def reverse(s)
```

Comment:
The time complexity isn't O(n^2), you are going to make recursive calls proportional to the size of the string.  So O(n)